raft topology: drop changing the raft voters config via storage_service

emdroid opened the pull request:
For the limited voters feature to work properly we need to make sure that we are only managing the voter status through the topology coordinator. This means that we should not change the node votership from the storage_service module for the raft topology directly.

We can drop the voter status changes from the storage_service module because the topology coordinator will handle the votership changes eventually. The calls in the storage_service module were not essential and were only used for optimization (improving the HA under certain conditions).
Furthermore, the other bundled commit improves the reaction again by reacting to the node `on_up()` and `on_down()` events, which again shortens the reaction time and improves the HA.

The change has effect on the timing in the tablets migration test though, as it previously relied on the node being made non-voter from the service_storage `raft_removenode()` function. The fix is to add another server to the topology to make sure we will keep the quorum.

Previously the test worked because the test waits for an injection to be reached and it was ensured that the injection (log line) has only been triggered after the node has been made non-voter from the `raft_removenode()`. This is not the case anymore. An alternative fix would be to wait for the first node to be made non-voter before stopping the second server, but this would make the test more complex (and it is not strictly required to only use 4 servers in the test, it has been only done for optimization purposes).

Fixes: scylladb/scylladb#22860

Refs: scylladb/scylladb#18793
Refs: scylladb/scylladb#21969

No backport: Part of the limited voters new feature, so this shouldn't to be backported.

patjed41 commented on service/topology_coordinator.cc:
```diff
@@ -2358,6 +2358,8 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                     // and then a coordinator again, it's fine - we'll just repeat this step.
                     // (If we're in `left` state when we try to restart we won't
                     // be able to become a voter - we'll be banned from the cluster.)
+                } else {
+                    co_await _group0.set_voter_status(node.id, can_vote::no, _as);
```
This is incorrect. Only decommissioned and replaced nodes reach the `left_token_ring` state. A removed node doesn't, it moves directly to `left` (it happens in `write_both_read_new`).

However, I think this change is unneeded. Funnily enough, it seems that we have three places where we make the node being removed a non-voter:
1. The one you are removing in `raft_removenode`.
2. On the topology coordinator, in the `write_both_read_old` handler:
```cpp
case topology::transition_state::write_both_read_old: {
    auto node = get_node_to_work_on(std::move(guard));

    ...

    if (_group0.is_member(node.id, true)) {
        // If we remove a node, we make it a non-voter early to improve availability in some situations.
        // There is no downside to it because the removed node is already considered dead by us.
        //
        // FIXME: removenode may be aborted and the already dead node can be resurrected. We should consider
        // restoring its voter state on the recovery path.
        if (node.rs->state == node_state::removing) {
            co_await _group0.set_voter_status(node.id, can_vote::no, _as);
        }
```
3. On the topology coordinator, in the `write_both_read_new` handler (removing from group 0 implies making a node non-voter):
```cpp
case node_state::removing: {
    co_await utils::get_local_injector().inject("delay_node_removal", [](auto& handler) {
        rtlogger.info("delay_node_removal: waiting for message");
        return handler.wait_for_message(db::timeout_clock::now() + std::chrono::minutes(5));
    });
    node = retake_node(co_await remove_from_group0(std::move(node.guard), node.id), node.id);
}
```

We have two simple options:
- keep 2 and 3,
- keep only 3.

The problem with keeping only 3 is that we will make the node being removed a non-voter later (streaming in `write_both_read_old` can take a lot of time). There may be other problems, I'm not sure.

The problem with keeping 2 and 3 is that we have to consider what happens when removenode is rolled back (we can roll back from `write_both_read_old` but we cannot roll back from `write_both_read_new` that is the topology operation is guaranteed to finish successfully if we reach `write_both_read_new`). This issue was described in the FIXME (code in 2). After moving all votership changes to the topology coordinator, this code in `rollback_current_topology_op` is for sure incorrect:
```cpp
case node_state::removing: {
    auto id = node.id;
    // The node was removed already. We need to add it back. Lets do it as non voter.
    // If it ever boots again it will make itself a voter.
    release_node(std::move(node));
    co_await _group0.group0_server().modify_config({raft::config_member{{id, {}}, false}}, {}, &_as);
    node = retake_node(co_await start_operation(), id);
}
```
The restarting node will never make itself a voter after your changes (it happened in `finish_setup_after_join`).

There is one thing I don't understand. To roll back removenode/decommission, we move the node back to the normal state. However, AFAIK, we also permanently ban a node being removed when we initiate removenode. Doing both of these things together doesn't make sense. @gleb-cloudius Can you explain?

We also have the code for decommissioned and replaced nodes in `write_both_read_old`:
```cpp
if (_group0.is_member(node.id, true)) {
    // If we remove a node, we make it a non-voter early to improve availability in some situations.
    // There is no downside to it because the removed node is already considered dead by us.
    //
    // FIXME: removenode may be aborted and the already dead node can be resurrected. We should consider
    // restoring its voter state on the recovery path.
    if (node.rs->state == node_state::removing) {
        co_await _group0.set_voter_status(node.id, can_vote::no, _as);
    }

    // If we decommission a node when the number of nodes is even, we make it a non-voter early.
    // All majorities containing this node will remain majorities when we make this node a non-voter
    // and remove it from the set because the required size of a majority decreases.
    //
    // FIXME: when a node restarts and notices it's a non-voter, it will become a voter again. If the
    // node restarts during a decommission, and we want the decommission to continue (e.g. because it's
    // at a finishing non-abortable step), we must ensure that the node doesn't become a voter.
    if (node.rs->state == node_state::decommissioning
            && raft::configuration::voter_count(_group0.group0_server().get_configuration().current) % 2 == 0) {
        if (node.id == _raft.id()) {
            rtlogger.info("coordinator is decommissioning and becomes a non-voter; "
                            "giving up leadership");
            co_await step_down_as_nonvoter();
        } else {
            co_await _group0.set_voter_status(node.id, can_vote::no, _as);
        }
    }
}
if (node.rs->state == node_state::replacing) {
    // We make a replaced node a non-voter early, just like a removed node.
    auto replaced_node_id = parse_replaced_node(node.req_param);
    if (_group0.is_member(replaced_node_id, true)) {
        co_await _group0.set_voter_status(replaced_node_id, can_vote::no, _as);
    }
}
```
I think we should keep the code for a replaced node if and only if we keep the code for a removed node. I'm not sure about the code for a decommissioning node. Maybe we should keep it but remove the `raft::configuration::voter_count(_group0.group0_server().get_configuration().current) % 2 == 0` condition. There is a different FIXME there, but moving all votership changes to the topology coordinator should fix it.

Of course, some of the things I'm writing about may be out of the scope of this PR.

emdroid replied:
Removed this change.

Kept both 2. and 3. from the above comment right now, but as discussed, actually looking into whether we could use the `on_down()` to detect the nodes that are down instead of the 2.

gleb-cloudius replied:
> There is one thing I don't understand. To roll back removenode/decommission, we move the node back to the normal state. However, I think we also permanently ban a node being removed when we initiate removenode. So, after rollback, we basically end up with a node that is normal but unable to contact other nodes (if it resurrects) and the only thing we can do is trying to remove/replace it again. @gleb-cloudius Is that right? I also remember there were some plans to implement unbanning. I'm trying to understand if a node should have a chance to become a voter after rollback.

When the code was written removenode did not ban node immediately. It was added later to unblock tablets operations. It is fine to have a blocked node in normal state as long as the node is dead. The only thing that can be done is removing it again or replacing. Decomissioning node should not be banned until the very end.

There were plans to have temporary banning to recover from unreachable nodes during topology operations. To unban such a anode it has to sync with raft state first. But those plans went nowhere.

emdroid replied:
Experimented with removing the voter from the `topology_coordinator::on_down()`, however it looks like it is problematic to call it from there, seems to be causing some kind of deadlock (the voter setting/unsetting code in raft group0 is calling read_barrier which seems to get stuck for some reason).

gleb-cloudius replied:
Only for removenode case.

patjed41 reviewed (changes requested):
Great to see this PR. Moving all voter changes to the topology coordinator can be considered a separate problem and is complicated enough to deserves its own PR(s).

emdroid commented on service/storage_service.cc:
```diff
@@ -6926,8 +6924,6 @@ future<join_node_request_result> storage_service::join_node_request_handler(join
         co_await utils::get_local_injector().inject("join-node-before-add-entry", utils::wait_for_message(5min));
 
         try {
-            // Make replaced node and ignored nodes non voters earlier for better HA
-            co_await _group0->set_voters_status(ignored_nodes_from_join_params(params), can_vote::no, _group0_as, raft_timeout{});
```
Updated the title and commit message to reflect all the changes.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15333/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15333/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3462/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6581/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14600/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15333/)
#### Failed Tests (2/40816):
- [test_simple_add_new_node_while_schema_changes[False]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15333/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest_with_topology_changes___test_simple_add_new_node_while_schema_changes_False_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes[False])
- [test_simple_add_new_node_while_schema_changes[False]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15333/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest_with_topology_changes___test_simple_add_new_node_while_schema_changes_False__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes[False])
#### Build Details:
- Duration: 6 hr 6 min
- Builder: spider7.cloudius-systems.com

emdroid commented:
^ rebased ^

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -143,6 +143,9 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
     std::optional<sstring> _rollback;
 
+    using group0_endpoint_voter_event = std::tuple<inet_address, can_vote>;
+    std::vector<group0_endpoint_voter_event> _group0_voter_events;
```
This should be hash since status may change twice before it is handled.

gleb-cloudius replied:
We do not need this at all. We can ask gossiper for the set of dead and alive nodes: get_live_members()/get_unreachable_members(). They already return host ids which will simplify the fiber as well.

gleb-cloudius replied:
We can calculate the set from current config and gossiper status.

gleb-cloudius replied:
> Yeah, could use a single bool

No even bool is needed. Event is triggerd -> recalculate -> if the set is empty do nothing.

emdroid replied:
Well, the bool would be to make sure we only react to the events originated from `on_up()` / `on_down()` (so to be checked in the `event.when()`) - there are some other places that could broadcast (so we wouldn't need to re-calculate at these points unnecessarily). Although it is a good question, maybe we could actually act on all topology event broadcasts.

gusev-p replied:
Do we actually need a fiber here? I guess we hit some deadlocks if we just call `set_voter_status` directly from `on_up/on_down`. We can put the call into an async lambda like in `storage_service::ip_address_updater`:
```
(void)futurize_invoke(ensure_alive([h = _ss._async_gate.hold()]() -> future<> {
  ///
})
```
pros:
  * no need to manage `_group0_voter_events` and the fiber;
  * `_topo_sm.event.broadcast();` causes a lot of other stuff to react, we can remove this overhead.

emdroid replied:
> Do we actually need a fiber here? I guess we hit some deadlocks if we just call `set_voter_status` directly from `on_up/on_down`. We can put the call into an async lambda like in `storage_service::ip_address_updater`:
> 
> ```
> (void)futurize_invoke(ensure_alive([h = _ss._async_gate.hold()]() -> future<> {
>   ///
> })
> ```

can give it a try (to see if it will still block in the read barrier)

> pros:
>     * no need to manage `_group0_voter_events` and the fiber;
>     * `_topo_sm.event.broadcast();` causes a lot of other stuff to react, we can remove this overhead.

however those will go away in any case, as we probably won't need the `_group0_voter_events` anyway ... and the `event.broadcast()` was called from the `on_up()` / `on_down()` anyway so it isn't a newly introduced broadcast

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +825,58 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                co_await _topo_sm.event.when([&] {
+                    return !_group0_voter_events.empty() || _as.abort_requested();
+                });
+                const auto& ep_to_id = co_await _sys_ks.load_host_ids();
```
Sigh. In the series I am working on I change on_up/on_down notifiers to get ids. IT would have made this code much simpler. Anyway to translate you need to use address map, not peers table.

emdroid replied:
Using the gossiper methods that give the host id directly now.

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +825,58 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+
+                for (const auto& [id, ep_can_vote] : events) {
+                    co_await _group0.set_voter_status(id, ep_can_vote, _as);
+                }
```
Why to you change it one by one? We need to add an API that can set/remove voters rights from a bunch of nodes atomically. Also this call here may hang in case we do not have a quorum any longer. Not sure if this is a problem.

gleb-cloudius replied:
May be we should detect this and do nothing but wait for the next event.

emdroid replied:
We actually have a way to set multiple nodes at once, but just as voters and non-voters separately, i.e. could separate voters and non-voters and do just 2 calls. However I'm wondering if that could actually make the possibility to lose the quorum worse...

Currently the `raft_group0::modify_raft_voter_status()` is indeed separate for voters and non-voters, although I suppose we could modify it to be able to handle both in a single call 🤔

gleb-cloudius replied:
> We actually have a way to set multiple nodes at once, but just as voters and non-voters separately, i.e. could separate voters and non-voters and do just 2 calls. However I'm wondering if that could actually make the possibility to lose the quorum worse...

I saw that but this API is not flexible enough. 

> 
> Currently the `raft_group0::modify_raft_voter_status()` is indeed separate for voters and non-voters, although I suppose we could modify it to be able to handle both in a single call 🤔

We should.

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +825,58 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                rtlogger.debug("group0 voters refresh fiber aborted");
+            } catch (seastar::abort_requested_exception) {
+                rtlogger.debug("group0 voters refresh fiber aborted");
+            } catch (group0_concurrent_modification&) {
```
So if this is thrown but `_group0_voter_events` is already cleared we lose events. This shows that we need to handle this like level triggering event and not edge triggering.

gusev-p commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +825,58 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                rtlogger.debug("group0 voters refresh fiber aborted");
+            } catch (group0_concurrent_modification&) {
+            } catch (term_changed_error&) {
+                rtlogger.debug("group0 voters refresh fiber notices term change {} -> {}", _term, _raft.get_current_term());
```
Why not break from the loop immediately on `request_aborted`, `abort_requested_exception` and `term_changed_error`? IIUC,  `_as` will be set eventually in these cases, but why rely on that?

emdroid replied:
Added the breaks - [diff](https://github.com/scylladb/scylladb/compare/2f60bd5204dc50fce8dc53247ee4aabd956a4476..00abcb8709be25ac7096518debb4b848e0f8f6aa)

In particular, this has been copied from the other fiber implementations - as Gleb mentioned, they have fairly similar structure, so we could think about unifying the template (although they are not exactly the same, so it might not be entirely trivial to unify them).

gusev-p commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +825,58 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                _group0_voter_events.clear();
+
+                for (const auto& [id, ep_can_vote] : events) {
+                    co_await _group0.set_voter_status(id, ep_can_vote, _as);
```
Should we log the decision `up/down -> change_voter_status` here?

emdroid replied:
Not sure if that is necessary - we will still see logs like these:
```
INFO  2025-02-19 18:58:40,972 [shard 0: gms] raft_group0 - making servers {78644cbe-33b9-452a-aeaa-3ad84f9c42f8} voters ...
INFO  2025-02-19 18:58:41,139 [shard 0: gms] raft_group0 - servers {78644cbe-33b9-452a-aeaa-3ad84f9c42f8} are now voters.
INFO  2025-02-19 18:58:43,026 [shard 0: gms] raft_group0 - making servers {bd9e9789-ef19-4737-976e-4ee6dbd41390} voters ...
INFO  2025-02-19 18:58:43,026 [shard 0: gms] raft_group0 - servers {bd9e9789-ef19-4737-976e-4ee6dbd41390} are now voters.
INFO  2025-02-19 18:59:05,427 [shard 0: gms] raft_group0 - making servers {78644cbe-33b9-452a-aeaa-3ad84f9c42f8} non-voters ...
INFO  2025-02-19 18:59:05,428 [shard 0: gms] raft_group0 - servers {78644cbe-33b9-452a-aeaa-3ad84f9c42f8} are now non-voters.
```
so we'll know there was a voter change (although indeed not strictly having the information that it has been a result of on_up/on_down).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15432/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15432/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_raft_no_quorum
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14688/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6671/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3549/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15432/)
#### Failed Tests (73/39557):
- [test_empty_reconciled_result](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_empty_reconciled_result) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result)
- [test_removenode_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest_with_topology_changes___test_removenode_rejoin) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removenode_rejoin)
- [test_empty_reconciled_result_with_paging](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_empty_reconciled_result_with_paging) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result_with_paging)
- [test_short_read_partitions](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_short_read_partitions) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_partitions)
- [test_repair_joint_row_3nodes_1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_joint_row_3nodes_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_joint_row_3nodes_1)
- [test_repair_same_row_diff_value_3nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_same_row_diff_value_3nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_same_row_diff_value_3nodes)
- [test_ttl_is_respected_on_delayed_replication](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/ttl_test/TestDistributedTTL/Tests___dtest_with_topology_changes___test_ttl_is_respected_on_delayed_replication) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ttl_is_respected_on_delayed_replication)
- [test_ttl_is_respected_on_repair[parallel_aggregates_disabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/ttl_test/TestDistributedTTL/Tests___dtest_with_topology_changes___test_ttl_is_respected_on_repair_parallel_aggregates_disabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ttl_is_respected_on_repair[parallel_aggregates_disabled])
- [test_ttl_is_respected_on_repair[parallel_aggregates_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/ttl_test/TestDistributedTTL/Tests___dtest_with_topology_changes___test_ttl_is_respected_on_repair_parallel_aggregates_enabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ttl_is_respected_on_repair[parallel_aggregates_enabled])
- [test_repair_joint_row_3nodes_2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15432/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_joint_row_3nodes_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_joint_row_3nodes_2)
#### Build Details:
- Duration: 10 hr
- Builder: i-0808c648ccb669b98 (m5ad.12xlarge)

emdroid commented:
^ rebased ^

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15455/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15455/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_raft_no_quorum
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14701/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6683/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3561/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15455/)
#### Failed Tests (63/40367):
- [test_empty_reconciled_result](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_empty_reconciled_result) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result)
- [test_empty_reconciled_result](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_tablets___test_empty_reconciled_result) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result)
- [test_empty_reconciled_result_with_paging](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_tablets___test_empty_reconciled_result_with_paging) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result_with_paging)
- [test_empty_reconciled_result_with_paging](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_empty_reconciled_result_with_paging) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result_with_paging)
- [test_short_read_partitions](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_tablets___test_short_read_partitions) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_partitions)
- [test_short_read_partitions](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_short_read_partitions) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_partitions)
- [test_hintedhandoff_dont_revive](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest_with_tablets___test_hintedhandoff_dont_revive) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_dont_revive)
- [test_basic_distributed](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/lwt_test/TestLwt/Tests___dtest_with_topology_changes___test_basic_distributed) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic_distributed)
- [test_no_streaming_on_second_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_no_streaming_on_second_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_streaming_on_second_repair)
- [test_no_streaming_on_second_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15455/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_tablets___test_no_streaming_on_second_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_streaming_on_second_repair)
#### Build Details:
- Duration: 5 hr 47 min
- Builder: spider5.cloudius-systems.com

gleb-cloudius commented on service/raft/raft_group0.cc:
```diff
@@ -586,13 +586,13 @@ future<> raft_group0::join_group0(std::vector<gms::inet_address> seeds, shared_p
         service::raft_group0& _group0;
         netw::messaging_service& _ms;
-        service::can_vote _can_vote;
+        bool _can_vote;
```
Why?

emdroid replied:
We in fact no longer really need the type for the voter setting methods (after the rework to allow setting voters and non-voters in a single call) ... but it's true that it is already there so we can keep it at least for the handshaker as it makes the calls a bit more clear. So reverted this removal.

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +822,52 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                rtlogger.debug("group0 voters refresh fiber aborted");
+            } catch (seastar::abort_requested_exception) {
+                rtlogger.debug("group0 voters refresh fiber aborted");
+            } catch (group0_concurrent_modification&) {
```
You need to retry without waiting on event again in this case.

emdroid replied:
Added the retry without waiting

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +822,52 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                    return raft::server_id{host_id.uuid()};
+                }) | std::ranges::views::filter([&group0_config](const auto& id) {
+                    // if the unreachable (dead) node is a voter, make it a non-voter
+                    return group0_config.can_vote(id);
```
We can define lambda `is_voter` and pass it here and `std::not_fn(is_voter)` above.  A little bit of funtional programming fun.

emdroid replied:
Good point, added the lambda

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -2174,15 +2220,6 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
-                    //
-                    // FIXME: removenode may be aborted and the already dead node can be resurrected. We should consider
-                    // restoring its voter state on the recovery path.
-                    if (node.rs->state == node_state::removing) {
```
Making replaced node non voter can also be dropped since the node that is been replaced has to be down.

emdroid replied:
Done

emdroid commented:
New version:
* changed the voters interface to allow setting both voters and non-voters in a single call
* simplified the `on_up()` / `on_down()` handling - not keeping the state, always re-taking the current gossiper state (that part will eventually go away with the final limited voters change anyway)

[diff](https://github.com/scylladb/scylladb/compare/5f0b734dac5733ef441b44000ec59ccf812d2fc3..2f60bd5204dc50fce8dc53247ee4aabd956a4476)

emdroid commented on service/storage_service.cc:
```diff
@@ -4118,8 +4118,6 @@ future<> storage_service::raft_removenode(locator::host_id host_id, locator::hos
         try {
-            // Make non voter during request submission for better HA
-            co_await _group0->set_voters_status(ignored_ids, can_vote::no, _group0_as, raft_timeout{});
             co_await _group0->client().add_entry(std::move(g0_cmd), std::move(guard), _group0_as, raft_timeout{});
```
Good point, added some details to the commit message (and the PR description).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15484/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15484/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_raft_no_quorum
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14729/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6708/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3587/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15484/)
#### Failed Tests (67/39605):
- [test_empty_reconciled_result](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_empty_reconciled_result) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result)
- [test_removenode_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest_with_topology_changes___test_removenode_rejoin) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removenode_rejoin)
- [test_empty_reconciled_result_with_paging](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_empty_reconciled_result_with_paging) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result_with_paging)
- [test_short_read_partitions](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_short_read_partitions) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_partitions)
- [test_changes_while_node_toggle](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/concurrent_schema_changes_test/TestConcurrentSchemaChanges/Tests___dtest_with_topology_changes___test_changes_while_node_toggle) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_changes_while_node_toggle)
- [test_repair_joint_row_3nodes_1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_joint_row_3nodes_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_joint_row_3nodes_1)
- [test_repair_same_row_diff_value_3nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_same_row_diff_value_3nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_same_row_diff_value_3nodes)
- [test_basic_distributed](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/lwt_test/TestLwt/Tests___dtest_with_topology_changes___test_basic_distributed) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic_distributed)
- [test_ttl_is_respected_on_repair[parallel_aggregates_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/ttl_test/TestDistributedTTL/Tests___dtest_with_topology_changes___test_ttl_is_respected_on_repair_parallel_aggregates_enabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ttl_is_respected_on_repair[parallel_aggregates_enabled])
- [test_ttl_is_respected_on_delayed_replication](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15484/testReport/junit/ttl_test/TestDistributedTTL/Tests___dtest_with_topology_changes___test_ttl_is_respected_on_delayed_replication) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ttl_is_respected_on_delayed_replication)
#### Build Details:
- Duration: 9 hr 52 min
- Builder: i-042a9b69b225b3691 (m5ad.12xlarge)

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +822,70 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                                            //
+                                            | std::ranges::to<std::unordered_set>();
+
+                co_await _group0.modify_voters(voters_to_add, voters_to_del, _as);
```
We need to be careful to not lose availability here. Lets consider a cluster with 2 DCs (dc1 and dc2) 3 nodes each.  One nodes goes down in each DC and we mark them as non voters. We have 4 voters now. Now two nodes that were down are going up and two voters die permanently before we manage to return a votership to two nodes that are now UP. Now we do not have a quorum permanently and need to recover.

Or we have 10 node cluster and we shutting it down one by one for whatever reason. When 2 nodes are left we can no longer remove a votership, so two nodes are remaining voters. Now we start the cluster again, but one of the voters cannot start for some reason -> need to recover.

May be both of those scenarios are rare  enough to accept that recover will be needed.

emdroid replied:
That is a good point. Maybe in case of a node just shutting down we shouldn't remove the votership at least under specific circumstances (in particular, if the voter count would drop under certain level, i.e. the limited voters count and we cannot distribute the voters to the other nodes that are still up) and only remove the voter in case the node is actually removed. Here (before the limited voters implementation) it is somewhat tricky, as all nodes are becoming voters. We could introduce some minimum, like e.g. 3 or 5 ... but this would further complicate this code that will be thrown away anyway when limited voters PR is done. So it might be better to remove the `on_up()` / `on_down()` handling from this PR altogether and move it to the final limited voters PR.

In particular, some tests that failed on this PR are actually exhibiting a quorum loss.

emdroid commented:
^ rebased, resolved conflicts ^

gusev-p commented on service/raft/raft_group0.cc:
```diff
@@ -858,31 +858,39 @@ future<> raft_group0::become_nonvoter(abort_source& as, std::optional<raft_timeo
     }
 
-    if (nodes.empty()) {
+    if (nodes_add.empty() && nodes_del.empty()) {
```
Can't we move this check to the beginning of the function? `group0_voter_refresher_fiber` reacts to all the changes in `_topo_sm.event`, a lot of them would be no-op-s, don't need `raft_upgrade_complete` each time.

emdroid replied:
Yes, that makes sense, moved.

gusev-p commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +822,70 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+
+                // wait for a moment to accumulate correlated changes
+                try {
+                    co_await seastar::sleep_abortable(std::chrono::seconds(1), _as);
```
Shouldn't we skip the sleep in case of `group0_concurrent_modification`? We can move the sleep into `if (!retry) {`.

Also `maybe_yield` at the end of the loop is not necessary, since we have already have other awaits.

emdroid replied:
Done.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15499/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15499/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_raft_no_quorum
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14739/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6716/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3596/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15499/)
#### Failed Tests (63/40302):
- [test_empty_reconciled_result](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_empty_reconciled_result) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result)
- [test_removenode_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest_with_topology_changes___test_removenode_rejoin) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removenode_rejoin)
- [test_empty_reconciled_result_with_paging](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_empty_reconciled_result_with_paging) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_empty_reconciled_result_with_paging)
- [test_short_read_partitions](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_short_read_partitions) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_partitions)
- [test_changes_while_node_toggle](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/concurrent_schema_changes_test/TestConcurrentSchemaChanges/Tests___dtest_with_topology_changes___test_changes_while_node_toggle) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_changes_while_node_toggle)
- [test_repair_joint_row_3nodes_2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_joint_row_3nodes_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_joint_row_3nodes_2)
- [test_repair_same_row_diff_value_3nodes_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_same_row_diff_value_3nodes_diff_shard_count) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_same_row_diff_value_3nodes_diff_shard_count)
- [test_basic_distributed](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/lwt_test/TestLwt/Tests___dtest_with_topology_changes___test_basic_distributed) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic_distributed)
- [test_repair_joint_row_3nodes_1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_joint_row_3nodes_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_joint_row_3nodes_1)
- [test_ttl_is_respected_on_delayed_replication](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15499/testReport/junit/ttl_test/TestDistributedTTL/Tests___dtest_with_topology_changes___test_ttl_is_respected_on_delayed_replication) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ttl_is_respected_on_delayed_replication)
#### Build Details:
- Duration: 6 hr 3 min
- Builder: spider5.cloudius-systems.com

emdroid commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +822,70 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+
+                // wait for a moment to accumulate correlated changes
+                try {
+                    co_await seastar::sleep_abortable(std::chrono::seconds(1), _as);
```
Good point, removed the extra try/catch.

emdroid commented on service/topology_coordinator.cc:
```diff
@@ -822,6 +822,70 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                // wait for a moment to accumulate correlated changes
+                try {
+                    co_await seastar::sleep_abortable(std::chrono::seconds(1), _as);
+                } catch (...) {
```
Shortened to 100ms for now.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15509/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15509/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_raft_no_quorum
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3603/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14746/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6723/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15509/)
#### Failed Tests (21/40296):
- [test_remove_failure_with_no_normal_token_owners_in_dc[True].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_remove_failure_with_no_normal_token_owners_in_dc_True__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_failure_with_no_normal_token_owners_in_dc[True].debug.1)
- [test_replace_with_no_normal_token_owners_in_dc[True].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_True__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc[True].debug.1)
- [test_remove_failure_with_no_normal_token_owners_in_dc[True].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_remove_failure_with_no_normal_token_owners_in_dc_True__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_failure_with_no_normal_token_owners_in_dc[True].debug.2)
- [test_replace_with_no_normal_token_owners_in_dc[True].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_True__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc[True].debug.2)
- [test_remove_failure_with_no_normal_token_owners_in_dc[True].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_remove_failure_with_no_normal_token_owners_in_dc_True__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_failure_with_no_normal_token_owners_in_dc[True].debug.3)
- [test_replace_with_no_normal_token_owners_in_dc[True].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_True__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc[True].debug.3)
- [test_not_enough_token_owners.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_not_enough_token_owners/Tests___Unit_Tests___test_not_enough_token_owners_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_not_enough_token_owners.dev.1)
- [test_not_enough_token_owners.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_not_enough_token_owners/Tests___Unit_Tests___test_not_enough_token_owners_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_not_enough_token_owners.dev.2)
- [test_not_enough_token_owners.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_not_enough_token_owners/Tests___Unit_Tests___test_not_enough_token_owners_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_not_enough_token_owners.dev.3)
- [test_remove_failure_with_no_normal_token_owners_in_dc[True].dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15509/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_remove_failure_with_no_normal_token_owners_in_dc_True__dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_failure_with_no_normal_token_owners_in_dc[True].dev.1)
#### Build Details:
- Duration: 6 hr 1 min
- Builder: spider5.cloudius-systems.com

emdroid commented:
^ rebased ^

emdroid commented:
New version:
* removing the `on_up()` / `on_down()` handling from the PR - will be moved to the final "limited voters" PR - seems it will need a more complex handling, so it doesn't make sense to do that here in a code that would be replaced anyway (and it also makes more sense because of the feature flag introduction - the feature flag to be the part of the whole final PR)

[diff](https://github.com/scylladb/scylladb/compare/2a1380f9693e8acfcf60e9cf16eabb98cb87ba94..2fb4f8e9c5c4fa222e3a463b60abced2625c9c11)

patjed41 commented on service/raft/raft_group0.cc:
```diff
@@ -858,31 +858,39 @@ future<> raft_group0::become_nonvoter(abort_source& as, std::optional<raft_timeo
-    group0_log.info("making servers {} {}...", nodes, status_str);
+    if (!nodes_add.empty()) {
+        group0_log.info("making servers {} voters ...", nodes_add);
+    }
+    if (!nodes_del.empty()) {
+        group0_log.info("making servers {} non-voters ...", nodes_del);
+    }
 
-    co_await modify_raft_voter_status(nodes, can_vote, as, timeout);
+    co_await modify_raft_voter_config(nodes_add, nodes_del, as, timeout);
 
-    group0_log.info("servers {} are now {}.", nodes, status_str);
+    if (!nodes_add.empty()) {
+        group0_log.info("servers {} are now voters.", nodes_add);
+    }
+    if (!nodes_del.empty()) {
+        group0_log.info("servers {} are now non-voters.", nodes_del);
+    }
```
Logging a few INFO logs per topology operation/shutdown/restart should be fine, but perhaps the voter registry would be a better place, I'm not sure. Let's at least ensure we don't log the same information in two places.

patjed41 replied:
OK

emdroid commented:
New version:
* removed the `test_raft_no_quorum.py` changes, will be done together with the main PR (as it makes more sense to do there because of being connected to the `on_up()` / `on_down()` changes)
* keeping the `modify_raft_voter_status()` name
* added the assertion for the empty voters/non-voters intersection

[diff](https://github.com/scylladb/scylladb/compare/2fb4f8e9c5c4fa222e3a463b60abced2625c9c11..ce812ebe516d9542c7cb70983f39450786ca3ea8)

emdroid commented on service/storage_service.cc:
```diff
@@ -4116,19 +4114,13 @@ future<> storage_service::raft_removenode(locator::host_id host_id, locator::hos
-    } else {
+    if (!error.empty()) {
         auto err = fmt::format("Removenode failed. See earlier errors ({}). Request ID: {}", error, request_id);
         rtlogger.error("{}", err);
         throw std::runtime_error(err);
     }
+
+    rtlogger.info("Removenode succeeded. Request ID: {}", request_id);
```
Done, moved to a separate commit

emdroid commented on test/topology_custom/test_raft_no_quorum.py:
```diff
@@ -15,8 +15,8 @@
-@pytest.fixture
-def raft_op_timeout(build_mode):
+@pytest.fixture(name="raft_op_timeout")  # avoid the W0621:redefined-outer-name pylint warning
+def fixture_raft_op_timeout(build_mode):
```
Will do in the main limited voters PR change

emdroid commented on service/raft/raft_group0.cc:
```diff
@@ -968,23 +976,28 @@ future<bool> raft_group0::wait_for_raft() {
-future<> raft_group0::modify_raft_voter_status(
-        const std::unordered_set<raft::server_id>& ids, can_vote can_vote, abort_source& as, std::optional<raft_timeout> timeout) {
-    co_await run_op_with_retry(as, [this, &ids, timeout, can_vote, &as]() -> future<operation_result> {
+future<> raft_group0::modify_raft_voter_config(const std::unordered_set<raft::server_id>& voters_add, const std::unordered_set<raft::server_id>& voters_del,
```
Updated to keep the `modify_raft_voter_status` previous name. For the parameter names I generally like the `voters_add` and `voters_del` as it is kind of uniform pattern and having `voters` / `non_voters` looks somewhat ugly to me 🤔

patjed41 replied:
I'm fine with `voters_add` and `voters_del`.

emdroid commented on service/raft/raft_group0.hh:
```diff
@@ -225,17 +225,22 @@ public:
+    //
+    // The `nodes_add` are changed to voters, the `nodes_del` are changed to non-voters.
+    //
+    // The `nodes_add` and `nodes_del` sets must not intersect.
```
Added the check.

emdroid replied:
I looked into the `std::set_intersection`, but apparently it only works for ordered ranges (i.e. the regular sorted sets, not the `std::unordered_set`) ... which is a pity indeed

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15560/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15560/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6760/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14784/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3642/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15560/)
#### Build Details:
- Duration: 5 hr 55 min
- Builder: spider8.cloudius-systems.com

patjed41 commented on service/raft/raft_group0.cc:
```diff
@@ -968,23 +976,28 @@ future<bool> raft_group0::wait_for_raft() {
         try {
             co_await _raft_gr.group0_with_timeouts().modify_config(std::move(add), {}, &as, timeout);
         } catch (const raft::commit_status_unknown& e) {
-            group0_log.info("modify_raft_voter_config({}): modify_config returned \"{}\", retrying", ids, e);
```
I still see "modify_raft_voter_config({}, {})" in the log and the function is called `modify_raft_voter_status`.

patjed41 commented on test/topology_custom/test_raft_no_quorum.py:
```diff
@@ -63,8 +63,8 @@ async def test_cannot_add_new_node(manager: ManagerClient, raft_op_timeout: int)
     servers += await manager.servers_add(servers_num=2)
 
     logger.info("stopping the last two nodes")
-    await asyncio.gather(manager.server_stop_gracefully(servers[2].server_id),
```
I'm resolving this comment since it's out of the scope of this PR bow, but please remember about it.

emdroid commented:
New version:
* fixed the log line

[diff](https://github.com/scylladb/scylladb/compare/ce812ebe516d9542c7cb70983f39450786ca3ea8..33fdd9a449cbf53ec1431552c08be7f2c2ecbc6e)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15578/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15578/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6770/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3652/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14794/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15578/)
#### Build Details:
- Duration: 6 hr 49 min
- Builder: i-081b42167a2e5c77d (m5d.16xlarge)

patjed41 commented:
@gleb-cloudius @gusev-p It seems that all open comments are related to the code that has been removed from this PR, but it would be better if looked at the new version.

I also can't queue it now because of the merge freeze.

gusev-p commented on service/raft/raft_group0.cc:
```diff
@@ -968,19 +986,24 @@ future<bool> raft_group0::wait_for_raft() {
-    co_await run_op_with_retry(as, [this, &ids, timeout, can_vote, &as]() -> future<operation_result> {
+future<> raft_group0::modify_raft_voter_status(const std::unordered_set<raft::server_id>& voters_add, const std::unordered_set<raft::server_id>& voters_del,
+        abort_source& as, std::optional<raft_timeout> timeout) {
+    co_await run_op_with_retry(as, [this, &voters_add, &voters_del, timeout, &as] -> future<operation_result> {
```
can't we just `return run_op_with_retry`? no need for `co_await`.

emdroid replied:
right, we should be able to do that (just need to be careful with the temporaries, in particular when moving the `std::vector<raft::config_member>` outside of the retry loop)

gusev-p commented on service/raft/raft_group0.cc:
```diff
@@ -858,31 +858,49 @@ future<> raft_group0::become_nonvoter(abort_source& as, std::optional<raft_timeo
     group0_log.info("becoming a non-voter (my id = {})...", my_id);
 
-    co_await modify_raft_voter_status({my_id}, can_vote::no, as, timeout);
+    co_await modify_raft_voter_status({}, {my_id}, as, timeout);
```
Why we need `modify_raft_voter_status`? We can use `modify_voters` instead of it in `become_nonvoter` (like we do in `make_nonvoter`) and inline it into `modify_voters` (simpler code structure).

emdroid replied:
There would be double logging in that case ... but anyway, after the limited voters we will actually not need the `raft_group0::become_nonvoter()` at all (the limited voters handler will not call it) and it can be removed after, so I'd not be too worried about it.

gusev-p commented on service/raft/raft_group0.cc:
```diff
@@ -968,19 +986,24 @@ future<bool> raft_group0::wait_for_raft() {
-            return raft::config_member{{id, {}}, static_cast<bool>(can_vote)};
+        add.reserve(voters_add.size() + voters_del.size());
+
+        std::transform(voters_add.begin(), voters_add.end(), std::back_inserter(add), [](raft::server_id id) {
```
We can do the `std::vector<raft::config_member>` calculation outside of the retry loop, it doesn't depend on it.

emdroid replied:
good point, done - one thing is that we then always need to copy it for the `modify_config()` instead of moving, but copying the final vector should still be better than re-creating it every time (although we might be pessimizing the first call that will succeed most of the time)

emdroid replied:
right, probably better to keep the previous version

emdroid commented:
New version:
* calculating the `std::vector<raft::config_member>` outside of the retry loop
* use direct return for `run_op_with_retry`

[diff](https://github.com/scylladb/scylladb/compare/c779caf7cac9378662cc793af82ea1b9b629a8b7..ce51b162220d4f59206a4efd79f5940272f84fa6)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15669/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15669/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3721/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6837/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14860/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15669/)
#### Build Details:
- Duration: 10 hr
- Builder: i-01ce9f1a335a64c71 (m5ad.12xlarge)

emdroid commented:
^ rebased ^

emdroid commented:
New version:
* fixed trailing whitespace

[diff](https://github.com/scylladb/scylladb/compare/91e3d9baf318e87a94441b2fe7a22895fbce3d42..8c67307971a2d2c47e16a518164717ef9d37ddb0)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15715/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15715/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets_migration
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6880/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3766/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14903/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15715/)
#### Build Details:
- Duration: 6 hr 8 min
- Builder: spider4.cloudius-systems.com
